### PR TITLE
feat/guest checkout

### PR DESCRIPTION
**Pull request by JakubSQ**
Guest

Opis:
- As a guest, I want to be able to browse the app, so I can view the product's offering without having to create an account.
- As a guest, I want to be able to add products to and manage my shopping cart, so I can prepare a list of products I want to buy without having to log in.
- As a guest, I want to be able to log in to my account during the checkout, so I can reuse one of my saved addresses or save a new one.
- As a guest, I want to be able to sign up during the checkout, so I can save my address for the next purchase.
- As a guest, I want to checkout without having an account, so I can finish my purchase quickly and easily.

Acceptance criteria:
- After clicking on the checkout button, the guest can decide not to log in or register and proceed without using an account
- There is a system & request spec

### CODE REVIEW

File: app/services/checkout/current_user.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+      end
+    end
+
+    private
+
+    def set_user(user, params)
+      @user = user
+      return @user if @user.present?
+
+      @user = User.where(email: params[:user_email], registered: false).first_or_initialize
```
Comment (kamilpuchala) [inline app/services/checkout/current_user.rb:19]: Pls check what will happen if you:
1. register as xxx@gmail.com
2. Log out
3. go to checkout
4. use xxx@gmail.com without login

Comment (kamilpuchala) [reply app/services/checkout/current_user.rb:19]: After create user without password(on checkout) pls check if you can login without password

File: app/services/checkout/creator.rb [hunk 2]
```diff
@@ -15,6 +18,7 @@ def call(cart, user, params)
     private
 
     def order_assignment(cart, user, params)
+      # return @error = user[:error] if user[:error].present?
```
Comment (kamilpuchala) [inline app/services/checkout/creator.rb:21]: ?

File: app/views/application/_navigation.html.erb [hunk 3]
```diff
@@ -14,6 +14,13 @@
       </ul>
       <ul class="navbar-nav ml-auto">
         <% if !user_signed_in? %>
+          <% if total_cart_items %>
+          <li class="nav-item <%= 'active' if request.path == cart_path(@cart) %>">
```
Comment (kamilpuchala) [inline app/views/application/_navigation.html.erb:18]: Decorator/presenter?

File: app/views/carts/show.html.erb [hunk 4]
```diff
@@ -21,9 +21,9 @@
 		<div class="col">
 		</div>
 		<div class="btn-group col mb-4" role="group" aria-label="Basic mixed styles example">
-				<%= link_to 'Empty cart', @cart, method: :delete, data: { confirm: "Are you sure?" }, class: "btn btn-danger" %>
-				<%= link_to 'Keep shopping', root_path, class: "btn btn-warning" %>
-				<%= link_to 'Checkout', new_order_path, class: "btn btn-success" %>	
+      <%= link_to 'Empty cart', @cart, method: :delete, data: { confirm: "Are you sure?" }, class: "btn btn-danger" %>
+      <%= link_to 'Keep shopping', root_path, class: "btn btn-warning" %>
+      <%= link_to 'Checkout', user_signed_in? ? new_order_path : new_user_session_path, class: "btn btn-success" %>
```
Comment (kamilpuchala) [inline app/views/carts/show.html.erb:26]: presenter?

File: app/views/devise/shared/_links.html.erb [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+<%- if devise_mapping.recoverable? && controller_name != 'passwords' && controller_name != 'registrations' %>
+  <%= link_to "Forgot your password?", new_password_path(resource_name) %><br />
+<% end %>
+
+<%- if devise_mapping.confirmable? && controller_name != 'confirmations' %>
+  <%= link_to "Didn't receive confirmation instructions?", new_confirmation_path(resource_name) %><br />
+<% end %>
+
+<%- if devise_mapping.lockable? && resource_class.unlock_strategy_enabled?(:email) && controller_name != 'unlocks' %>
```
Comment (kamilpuchala) [inline app/views/devise/shared/_links.html.erb:17]: pls check the differences between `<%-`, vs `<%`

**Comment by kamilpuchala** [comment]
pls clean devise views

File: app/services/checkout/current_user.rb (hunk 1 above)
Comment (JakubSQ) [reply app/services/checkout/current_user.rb:19]: > After create user without password(on checkout) pls check if you can login without password

User cannot log in without password. Devise reject query

Comment (JakubSQ) [reply app/services/checkout/current_user.rb:19]: > Pls check what will happen if you:
> 
> 1. register as [xxx@gmail.com](mailto:xxx@gmail.com)
> 2. Log out
> 3. go to checkout
> 4. use [xxx@gmail.com](mailto:xxx@gmail.com) without login

You are right! Query raised an error "Email has already been taken". Solution: I deleted 'registered: false' part in where clause.

File: app/services/checkout/creator.rb (hunk 2 above)
Comment (JakubSQ) [reply app/services/checkout/creator.rb:21]: trash ;)

File: app/views/application/_navigation.html.erb (hunk 3 above)
Comment (JakubSQ) [reply app/views/application/_navigation.html.erb]: PathPresenter added

File: app/views/carts/show.html.erb (hunk 4 above)
Comment (JakubSQ) [reply app/views/carts/show.html.erb]: PathPresenter added

File: app/views/devise/shared/_links.html.erb (hunk 5 above)
Comment (JakubSQ) [reply app/views/devise/shared/_links.html.erb:17]: <%- Avoids line break before expression.

**Pull request merged**